File: src/services/worktree.ts
import { invoke } from "@tauri-apps/api/core";

export interface WorktreeInfo {
  path: string;
  branch: string | null;
  head_oid: string | null;
  head_short_oid: string | null;
  is_main: boolean;
  is_clean: boolean;
}

export function listWorktrees(tabId: string): Promise<WorktreeInfo[]> {
  return invoke<WorktreeInfo[]>("list_worktrees", { tabId });
}

export function addWorktree(
  tabId: string,
  path: string,
  branch: string,
): Promise<void> {
  return invoke<void>("add_worktree", { tabId, path, branch });
}

export function removeWorktree(tabId: string, path: string): Promise<void> {
  return invoke<void>("remove_worktree", { tabId, path });
}

File: src-tauri/src/commands/worktree.rs
use tauri::State;

use crate::commands::with_repo;
use crate::git::types::WorktreeInfo;
use crate::state::AppState;

#[tauri::command]
pub fn list_worktrees(
    tab_id: String,
    state: State<'_, AppState>,
) -> Result<Vec<WorktreeInfo>, String> {
    with_repo(&state, &tab_id, |backend| {
        backend.list_worktrees().map_err(|e| e.to_string())
    })
}

#[tauri::command]
pub fn add_worktree(
    tab_id: String,
    path: String,
    branch: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    with_repo(&state, &tab_id, |backend| {
        backend
            .add_worktree(&path, &branch)
            .map_err(|e| e.to_string())
    })
}

#[tauri::command]
pub fn remove_worktree(
    tab_id: String,
    path: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    with_repo(&state, &tab_id, |backend| {
        backend.remove_worktree(&path).map_err(|e| e.to_string())
    })
}
